File: dotnet/src/Core/MessageAclChecker.cs
namespace Agience.Core
{
    public class MessageAclChecker
    {
        public const string EVENT_PREFIX = "event/";
        public const string CONNECT_PREFIX = "connect/";

        private const string ALL = "0";
        private const string NONE = "-";
        private const string ANY_EXCLUSIVE = "*";
        private const string ANY_INCLUSIVE = "+";
        private const string QUERY = "?";

        private const int READ = 1;
        private const int WRITE = 2;
        private const int READ_WRITE = 3;
        private const int SUBSCRIBE = 4;

        private Func<string, string, string, Task<bool>> _verifyHostSourceTargetRelationships;

        public MessageAclChecker(Func<string, string, string, Task<bool>> verifyHostSourceTargetRelationships)
        {
            _verifyHostSourceTargetRelationships = verifyHostSourceTargetRelationships;
        }

        // TODO: Rework and document all of this

        public async Task<bool> CheckAccessControl(AclCheckRequest? aclRequest, string? hostId, List<string> roles, string? authorityId)
        {
            // TODO: We'll need to lock this down
            if (aclRequest.topic.StartsWith(CONNECT_PREFIX))
            {
                return true;
            }

            // TODO: This just makes it compatible with the current implementation. We should remove this.
            if (aclRequest.topic.StartsWith(EVENT_PREFIX))
            {
                aclRequest.topic = aclRequest.topic.Remove(0, EVENT_PREFIX.Length);
            }

            if (aclRequest == null || string.IsNullOrWhiteSpace(aclRequest.topic) || aclRequest.acc == 0)
            {
                return false;
            }

            if (aclRequest.acc == READ_WRITE)
            {
                return false;
            }

            var masks = GetUserMasks(aclRequest.acc, roles, authorityId, hostId);

            if (await IsTopicAllowed(aclRequest.topic, masks, hostId, aclRequest.acc))
            {
                return true;
            }

            return false;
        }

        private List<string> GetUserMasks(int accessType, List<string> roles, string? authorityId, string? hostId)
        {
            bool isAuthority = roles.Contains("authority");
            bool isHost = roles.Contains("host");

            List<string> masks = new();

            if (isAuthority && authorityId != null)
            {
                if (accessType == READ || accessType == SUBSCRIBE)
                {
                    masks.Add($"{ANY_EXCLUSIVE}/{authorityId}/{NONE}/{NONE}"); // Any -> Authority
                }
                if (accessType == WRITE)
                {
                    masks.Add($"{NONE}/{authorityId}/{ANY_INCLUSIVE}/{ANY_INCLUSIVE}"); // Authority -> Any Host or Agent                    
                }
            }

            if (isHost && hostId != null)
            {
                if (accessType == READ || accessType == SUBSCRIBE)
                {
                    masks.Add($"{NONE}/{authorityId}/{ALL}/{NONE}"); // Authority -> All Hosts
                    masks.Add($"{NONE}/{authorityId}/{hostId}/{NONE}"); // Authority -> Host
                    masks.Add($"{NONE}/{authorityId}/{hostId}/{QUERY}"); // Authority -> Agent
                    masks.Add($"{ANY_EXCLUSIVE}/{authorityId}/{NONE}/{QUERY}"); // Any -> Agent
                }
                if (accessType == WRITE)
                {
                    masks.Add($"{hostId}/{authorityId}/{NONE}/{NONE}"); // Host -> Authority
                    masks.Add($"{QUERY}/{authorityId}/{NONE}/{NONE}"); // Agent -> Authority
                    masks.Add($"{QUERY}/{authorityId}/{NONE}/{QUERY}"); // Agent -> Agent

                }
            }
            return masks;
        }

        private async Task<bool> IsTopicAllowed(string topic, List<string> masks, string? hostId, int accessType)
        {
            foreach (var mask in masks)
            {
                if (CheckMask(topic, mask, accessType))
                {
                    return true;
                }

                if (mask.Contains(QUERY)) // TODO: Subject to "?" injection attack.
                {
                    if (hostId == null) throw new ArgumentNullException(nameof(hostId));

                    if (await CheckQueryMaskAsync(topic, mask, hostId, accessType))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        private async Task<bool> CheckQueryMaskAsync(string topic, string mask, string hostId, int accessType)
        {
            var topicParts = topic.Split('/');
            var maskParts = mask.Split('/');

            if (!IsValidTopicAndMask(topicParts, maskParts)) { return false; }

            var sourceId = maskParts[0] == QUERY ? topicParts[0] : null;
            var targetAgentId = maskParts[3] == QUERY ? topicParts[3] : null;

            if (sourceId == NONE || targetAgentId == NONE) { return false; }

            if (accessType == WRITE && sourceId == targetAgentId)
            {
                return false; // Can't send to self
            }

            return await _verifyHostSourceTargetRelationships(hostId, sourceId, targetAgentId);
        }

        private bool CheckMask(string topic, string mask, int accessType)
        {
            var topicParts = topic.Split('/');
            var maskParts = mask.Split('/');

            if (!IsValidTopicAndMask(topicParts, maskParts)) return false;

            if (topicParts[0] == null || maskParts[0] == null) { return false; }

            // First part is the sender id. Read from any sender. Otherwise, the sender id must match the claims.
            if (accessType != READ)
            {
                if (accessType == SUBSCRIBE && topicParts[0] != ANY_INCLUSIVE) { return false; }
                if (accessType == WRITE && topicParts[0] != maskParts[0]) { return false; }
            }

            for (int i = 1; i < maskParts.Length; i++)
            {
                switch (maskParts[i])
                {
                    case ANY_INCLUSIVE:
                        continue;
                    case ANY_EXCLUSIVE when topicParts[i] != ALL:
                        continue;
                    case ALL when topicParts[i] == ALL:
                        continue;
                    case NONE when topicParts[i] == NONE:
                        continue;
                    case var m when m != topicParts[i]:
                        return false;
                }
            }
            return true;
        }

        private bool IsValidTopicAndMask(string[] topicParts, string[] maskParts)
        {
            return topicParts.Length == 4 && maskParts.Length == 4;
        }

        public class AclCheckRequest
        {
            public int acc { get; set; }
            public string clientid { get; set; } = string.Empty;
            public string topic { get; set; } = string.Empty;
        }

        // For logging
        private string PrintConst(int value)
        {
            return value switch
            {
                0 => "0",
                1 => nameof(READ),
                2 => nameof(WRITE),
                3 => nameof(READ_WRITE),
                4 => nameof(SUBSCRIBE),
                _ => "UNKNOWN"
            };
        }
    }
}

File: core/sdk/python/message_acl_checker.py
from typing import List, Optional, Callable, Awaitable
from pydantic import BaseModel
from functools import partial


class AclCheckRequest(BaseModel):
    acc: int
    clientid: str = ""
    topic: str = ""


class MessageAclChecker:
    EVENT_PREFIX: str = "event/"
    CONNECT_PREFIX: str = "connect/"

    # Constants for access control
    ALL: str = "0"
    NONE: str = "-"
    ANY_EXCLUSIVE: str = "*"
    ANY_INCLUSIVE: str = "+"
    QUERY: str = "?"

    # Access types
    READ: int = 1
    WRITE: int = 2
    READ_WRITE: int = 3
    SUBSCRIBE: int = 4

    def __init__(self, verify_host_source_target_relationships: Callable[[str, str, str], Awaitable[bool]]):
        self._verify_host_source_target_relationships = verify_host_source_target_relationships

    async def check_access_control(
        self,
        acl_request: Optional[AclCheckRequest],
        host_id: Optional[str],
        roles: List[str],
        authority_id: Optional[str]
    ) -> bool:
        # Handle connection prefixes
        if acl_request and acl_request.topic.startswith(self.CONNECT_PREFIX):
            return True

        # Remove event prefix if present
        if acl_request and acl_request.topic.startswith(self.EVENT_PREFIX):
            acl_request.topic = acl_request.topic[len(self.EVENT_PREFIX):]

        # Basic validation
        if (not acl_request or
            not acl_request.topic or
            acl_request.topic.isspace() or
                acl_request.acc == 0):
            return False

        if acl_request.acc == self.READ_WRITE:
            return False

        masks = self._get_user_masks(
            acl_request.acc, roles, authority_id, host_id)

        return await self._is_topic_allowed(acl_request.topic, masks, host_id, acl_request.acc)

    def _get_user_masks(
        self,
        access_type: int,
        roles: List[str],
        authority_id: Optional[str],
        host_id: Optional[str]
    ) -> List[str]:
        is_authority = "authority" in roles
        is_host = "host" in roles
        masks = []

        if is_authority and authority_id:
            if access_type in (self.READ, self.SUBSCRIBE):
                # Any -> Authority
                masks.append(
                    f"{self.ANY_EXCLUSIVE}/{authority_id}/{self.NONE}/{self.NONE}")

            if access_type == self.WRITE:
                # Authority -> Any Host or Agent
                masks.append(
                    f"{self.NONE}/{authority_id}/{self.ANY_INCLUSIVE}/{self.ANY_INCLUSIVE}")

        if is_host and host_id:
            if access_type in (self.READ, self.SUBSCRIBE):
                masks.extend([
                    # Authority -> All Hosts
                    f"{self.NONE}/{authority_id}/{self.ALL}/{self.NONE}",
                    # Authority -> Host
                    f"{self.NONE}/{authority_id}/{host_id}/{self.NONE}",
                    # Authority -> Agent
                    f"{self.NONE}/{authority_id}/{host_id}/{self.QUERY}",
                    # Any -> Agent
                    f"{self.ANY_EXCLUSIVE}/{authority_id}/{self.NONE}/{self.QUERY}"
                ])

            if access_type == self.WRITE:
                masks.extend([
                    # Host -> Authority
                    f"{host_id}/{authority_id}/{self.NONE}/{self.NONE}",
                    # Agent -> Authority
                    f"{self.QUERY}/{authority_id}/{self.NONE}/{self.NONE}",
                    # Agent -> Agent
                    f"{self.QUERY}/{authority_id}/{self.NONE}/{self.QUERY}"
                ])

        return masks

    async def _is_topic_allowed(
        self,
        topic: str,
        masks: List[str],
        host_id: Optional[str],
        access_type: int
    ) -> bool:
        for mask in masks:
            if self._check_mask(topic, mask, access_type):
                return True

            if self.QUERY in mask:  # TODO: Subject to "?" injection attack
                if not host_id:
                    raise ValueError(
                        "host_id cannot be None when processing query masks")

                if await self._check_query_mask_async(topic, mask, host_id, access_type):
                    return True

        return False

    async def _check_query_mask_async(
        self,
        topic: str,
        mask: str,
        host_id: str,
        access_type: int
    ) -> bool:
        topic_parts = topic.split('/')
        mask_parts = mask.split('/')

        if not self._is_valid_topic_and_mask(topic_parts, mask_parts):
            return False

        source_id = topic_parts[0] if mask_parts[0] == self.QUERY else None
        target_agent_id = topic_parts[3] if mask_parts[3] == self.QUERY else None

        if source_id == self.NONE or target_agent_id == self.NONE:
            return False

        if access_type == self.WRITE and source_id == target_agent_id:
            return False  # Can't send to self

        return await self._verify_host_source_target_relationships(host_id, source_id, target_agent_id)

    def _check_mask(self, topic: str, mask: str, access_type: int) -> bool:
        topic_parts = topic.split('/')
        mask_parts = mask.split('/')

        if not self._is_valid_topic_and_mask(topic_parts, mask_parts):
            return False

        if not topic_parts[0] or not mask_parts[0]:
            return False

        # First part is the sender id. Read from any sender. Otherwise, the sender id must match the claims.
        if access_type != self.READ:
            if access_type == self.SUBSCRIBE and topic_parts[0] != self.ANY_INCLUSIVE:
                return False
            if access_type == self.WRITE and topic_parts[0] != mask_parts[0]:
                return False

        for i in range(1, len(mask_parts)):
            mask_part = mask_parts[i]
            topic_part = topic_parts[i]

            if mask_part == self.ANY_INCLUSIVE:
                continue
            elif mask_part == self.ANY_EXCLUSIVE and topic_part != self.ALL:
                continue
            elif mask_part == self.ALL and topic_part == self.ALL:
                continue
            elif mask_part == self.NONE and topic_part == self.NONE:
                continue
            elif mask_part != topic_part:
                return False

        return True

    def _is_valid_topic_and_mask(self, topic_parts: List[str], mask_parts: List[str]) -> bool:
        return len(topic_parts) == 4 and len(mask_parts) == 4

    @staticmethod
    def _print_const(value: int) -> str:
        const_map = {
            0: "0",
            1: "READ",
            2: "WRITE",
            3: "READ_WRITE",
            4: "SUBSCRIBE"
        }
        return const_map.get(value, "UNKNOWN")
